src/Common/ErrorCodes.cpp

```diff
@@ -610,6 +611,23 @@ namespace ErrorCodes
     constexpr ErrorCode END = 1002;
     ErrorPairHolder values[END + 1]{};
 
+    std::set retryable_errors = {
```

Comment:
Maybe we can just mark each error in-place (in definition)? E.g. 
```
   M(42, NUMBER_OF_ARGUMENTS_DOESNT_MATCH, /*is_retryable=*/false) \
   M(43, ILLEGAL_TYPE_OF_ARGUMENT, /*is_retryable=*/false) \
   M(44, ILLEGAL_COLUMN, /*is_retryable=*/false) \
```

Reply:
Some of these are not retriable though: `TIMEOUT_EXCEEDED`, `TABLE_IS_READ_ONLY`, `TOO_MANY_PARTS`, `KEEPER_EXCEPTION` for example  might refer to a problem when inserted into a materialized view and part of the data has already being ingested. Retrying might duplicate data.

Reply:
>  TIMEOUT_EXCEEDED, TABLE_IS_READ_ONLY, TOO_MANY_PARTS, KEEPER_EXCEPTION for example might refer to a problem when inserted into a materialized view

Maybe we should return an exception related to MV, then?

Reply:
Well, the same can happen without MVs too: You start inserting blocks, keeper dies, you get KEEPER_EXCEPTION for the next block in the query (some have been ingested and some haven't). Retrying might or might not duplicate data depending on the query settings and table definition (materialized column with non deterministic column for example).

Without transactions I don't think there is a simple way to say that an INSERT is retriable.

Reply:
Let's start with this list and then refine...

Reply:
> Maybe we can just mark each error in-place (in definition)?

+

Reply:
> Maybe we can just mark each error in-place (in definition)?

sure, I will update this draft accordingly, it was created mainly to discuss the naming and the error codes.

Reply:
> Well, the same can happen without MVs too: You start inserting blocks, keeper dies, you get KEEPER_EXCEPTION for the next block in the query (some have been ingested and some haven't). Retrying might or might not duplicate data depending on the query settings and table definition (materialized column with non deterministic column for example).
> 
> Without transactions I don't think there is a simple way to say that an INSERT is retriable.

That's totally fine, and it's expected behavior. You should always expect errors like these, and you should use deduplication when retrying.

Reply:
> That's totally fine, and it's expected behavior. You should always expect errors like these, and you should use deduplication when retrying.

It's expected behaviour, but that doesn't mean it's safe to retry. For example, it's common to add a new column to the data to set the insert time (via materialized column or via materialized view with now()) and this will break deduplication. With MVs it's even worse because you get JOINs and other non deterministic problems.

I'd rather be conservative and accurate when giving the go-ahead to retry a query. Better a false negative than a false positive. One alternative is to consider an INSERT is retriable if:
* Deduplication is on. This might be obvious, but it's not always the case.
* Only the landing table was used, of it materialized views were used they don't include other tables. 
* No non-deterministic functions were used.

We are already collecting the information about used tables and functions, so it seems possible. There might be more corner cases to consider.

Reply:
> With MVs it's even worse because you get JOINs and other non deterministic problems.

I haven't said anything. Consider joinGet / dictGet. Same exact problem as MV with JOINs.

Reply:
@Algunenano @tavplubix @alexey-milovidov 
I'd like to kindly ask you to provide "safe" retryable error codes you agree on so we will just keep these for now and finalize the PR.

Reply:
I'm afraid that the list of totally safe retriable error codes will be empty (well, maybe it will contain just a few very specific codes)

Reply:
+1
A side question: if there are errors that are retriable, why isn't CH retrying them itself?

Reply:
> maybe it will contain just a few very specific codes

that's a start already :)

Reply:
> I'm afraid that the list of totally safe retriable error codes will be empty (well, maybe it will contain just a few very specific codes)

Let's step back and look from the customer's point of view. What would be our recommendations to a customer who received one of the exceptions? Most error messages aren't actionable and most of the customers might be not familiar with the underlying mechanics of ClickHouse (like MV).

> One alternative is to consider an INSERT is retriable if:
> - Deduplication is on. This might be obvious, but it's not always the case.
> - Only the landing table was used, of it materialized views were used they don't include other tables.
> - No non-deterministic functions were used.

I agree, it's a more sound approach. If we decide to go this way, we need someone with good expertise in the code base to help.

Reply:
Well, there's no silver bullet. My personal recommendation is to always retry with [exponential] backoff (having backoff is extremely important) if the error looks like retriable one. The number of retries must be limited. 

What may happen if you retry a non-retriable error:
 - If it's an insert query, you may get data duplication. But it's not a problem if you use deduplication (however, deduplication may be tricky in some cases, but it comes to some "retriable" errors too)
 - You will just get the same (or a similar) error again on retry. That's why the number of retries must be limited. For example, `TABLE_IS_READ_ONLY` can mean that the table is temporarily readonly while it reconnects to Keeper, in this case you can safely retry. But it also can mean that something is completely wrong and the table doesn't have metadata in Keeper (or it's broken), and in this case it's permanently readonly. It's still safe to retry, but retries will never succeed
 - Retries may amplify some non-retriable errors, and it may basically cause an unintentional DoS attack (the worst case). That's why backoff is strictly required. For example, even if `TIMEOUT_EXCEEDED` and similar errors may look like retriable errors, they also can mean that the request is too heavy, and just retrying again and again without increasing the timeout can overload the system and make things worse. 

Sometimes it makes sense to have different retry policies for different kinds of errors and retry some of them more aggressively and others more conservatively. 

I think it's easier to start with a list of non-retriable errors. For example, `SYNTAX_ERROR`, `CANNOT_PARSE_*` and `BAD_ARGUMENTS` are definitely not retriable. Some errors like `UNKNOWN_TABLE` and `UNKNOWN_FUNCTION` may be retriable or not depending on your workload. For example, if you constantly create/drop tables/udfs asynchronously then you can get these errors as a result of eventually consistent replication of DDLs and UDFs. And if you don't do this, then these errors are definitely not retriable. 

> most of the customers might be not familiar with the underlying mechanics of ClickHouse (like MV)

If you aren't familiar with MV you don't use MV. If you use MV, then you have to understand what you use